hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/DigestSaslClientAuthenticationProvider.java

```diff
@@ -0,0 +1,118 @@
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.Map;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.RealmCallback;
+import javax.security.sasl.RealmChoiceCallback;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslClient;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.security.SaslUtil;
+import org.apache.hadoop.hbase.security.SecurityInfo;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.yetus.audience.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.hbase.shaded.protobuf.generated.RPCProtos.UserInformation;
+
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.AUTHENTICATION)
+@InterfaceStability.Evolving
+public class DigestSaslClientAuthenticationProvider extends DigestSaslAuthenticationProvider
+    implements SaslClientAuthenticationProvider {
+
+  private static final String MECHANISM = "DIGEST-MD5";
+
+  @Override
+  public SaslClient createClient(Configuration conf, InetAddress serverAddr,
+      SecurityInfo securityInfo, Token<? extends TokenIdentifier> token, boolean fallbackAllowed,
+      Map<String, String> saslProps) throws IOException {
+    return Sasl.createSaslClient(new String[] { MECHANISM }, null, null,
+        SaslUtil.SASL_DEFAULT_REALM, saslProps, new DigestSaslClientCallbackHandler(token));
+  }
+
+  public static class DigestSaslClientCallbackHandler implements CallbackHandler {
+    private static final Logger LOG =
+        LoggerFactory.getLogger(DigestSaslClientCallbackHandler.class);
+    private final String userName;
+    private final char[] userPassword;
+
+    public DigestSaslClientCallbackHandler(Token<? extends TokenIdentifier> token) {
+      this.userName = SaslUtil.encodeIdentifier(token.getIdentifier());
+      this.userPassword = SaslUtil.encodePassword(token.getPassword());
+    }
+
+    @Override
+    public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+      NameCallback nc = null;
+      PasswordCallback pc = null;
+      RealmCallback rc = null;
+      for (Callback callback : callbacks) {
+        if (callback instanceof RealmChoiceCallback) {
+          continue;
+        } else if (callback instanceof NameCallback) {
+          nc = (NameCallback) callback;
+        } else if (callback instanceof PasswordCallback) {
+          pc = (PasswordCallback) callback;
+        } else if (callback instanceof RealmCallback) {
```

Comment:
Digest mechanism shall not dealt with realm callback, iirc.

Reply:
This was just a copy-past from the old implementation. Can we do clean-up separately?

Reply:
n/m, it already says it's ignored, let me clean it up :)

Reply:
According to https://tools.ietf.org/html/rfc2831#page-18, the `RealmCallback` can be used by a server, but we don't use it in HBase. I'll drop it client side, but leave a big-fat comment to why it's not there.

Reply:
Turns out dropping things actually doesn't work. Will comment with what I find.